### PR TITLE
Separate Algorithms for GE0 and GE1/1 GE2/1 Segments

### PR DESCRIPTION
**Pull request by watson-ij**
#### PR description:

For ME0, the RU segment builder was default while GEMSegmentAlgorithm is used for the other GEM stations. This PR allows separate GE0 and other GEM algorithms, and defaults GE0 to use the RU algorithm, and GE1/1 / GE2/1 to use the GEMSegmentAlgorithm, as is the default for the ME0/GEM setup.

@jshlee

#### PR validation:

We checked the performance of the GE0 RU algorithm and found it to work as for ME0. I also ran runTheMatrix with workflow 34621.0 and confirmed that both station 0 and 1/2 segments are being created properly.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by srimanob** [comment]
Please test

**Comment by srimanob** [comment]
@watson-ij 
Could you please provide material or plots as a result of this PR? For example, what do you expect to see when replacing GEMSegmentAlgorithm (current) with RU (introduce in this PR)? Will it be spotted using DQM?

This is to complete the documentation for the PR. Thanks.

**Comment by watson-ij** [comment]
@seulgi324, do you have a link to a presentation showing the difference?

**Comment by watson-ij** [comment]
> @watson-ij
> Could you please provide material or plots as a result of this PR? For example, what do you expect to see when replacing GEMSegmentAlgorithm (current) with RU (introduce in this PR)? Will it be spotted using DQM?
> 
> This is to complete the documentation for the PR. Thanks.

GEMSegments aren't currently tracked by DQM, so they won't show up directly, but it could have an impact on the muons. We saw that the default algorithm produces a lot of segments with only a few rechits, whereas the RU algorithm requires at least 4. The RU algorithm is also finding more segments with 6 rechits as is nominal for ME0.

File: RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc [hunk 1]
```diff
@@ -11,15 +11,20 @@
...
+                                        << " name: " << segAlgoName;
 
   // SegAlgo parameter set
   segAlgoPSet = ps.getParameter<edm::ParameterSet>("algo_pset");
+  ge0AlgoPSet = ps.getParameter<edm::ParameterSet>("ge0_pset");
 
   // Ask factory to build this algorithm, giving it appropriate ParameterSet
-  algo = GEMSegmentBuilderPluginFactory::get()->create(algoName, segAlgoPSet);
+  segAlgo = GEMSegmentBuilderPluginFactory::get()->create(segAlgoName, segAlgoPSet);
+  // Ask factory to build this algorithm, giving it appropriate ParameterSet
```
Comment (perrotta) [inline RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc:26]: Having twice the same comment is probably redundant...
If you remove one, then in the other one: "this algorithm, giving it" -> "these algorithms, giving them"

Comment (watson-ij) [reply RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc]: Yep, updated

**Comment by srimanob** [comment]
Please test

**Comment by watson-ij** [comment]
> Could you please provide material or plots as a result of this PR?

To complete this: @seulgi324 gave a  talk today showing the rechit distribution differences between the default and RU algorithms, in particular, showing the increase in 6-hit segments I mentioned earlier.

https://indico.cern.ch/event/1010727/contributions/4241679/attachments/2195182/3711306/Feb%2023rd%202021_ME0toGE0_Seulgi%20Kim.pdf

**Comment by perrotta** [comment]
The RU segment builder is implemented for the GE0, instead of the GEMSegmentAlgorithm which is used for all other GEM stations. The RU segment builder is what was originally implemented for ME0, and the GEMSegmentAlgorithm replaced it for GE0 in previous #31480: but it turned out to be sub-optimal for the GE0 station (something which was already possibly anticipated in the comments of #31480).

In the presentation at the GEM DPG linked here differences (expected) are reported for the number of segments and recHits. These differences cannot likely be noticed in the low stat Phase2 validations available in the automatic PR tests because of the low statistics: it looks like that they are however endorsed by the GEM DPG, something which was already possibly anticipated in the comments of #31480.

As such, the PR could be considered suitable for being integrated in CMSSW.

At page 6 of the same presentation there is however a mention about "an issue with ME0": could you please elaborate further about it, and specify with some more detail what such an "issue" is, and whether you plan/need to fix it already in this PR?

**Comment by watson-ij** [comment]
> At page 6 of the same presentation there is however a note about "an issue with ME0": could you please elaborate further about it, and specify with some more detail what such an "issue" is, and whether you plan/need to fix it already in this PR?

@perrotta, the issue referred to there was that the ME0 RecHit algorithm never implemented clustering, so that we have a rechit for every digi rather than clustering the contiguous digis into a single rechit, causing duplicate segments. This is already fixed for GE0 by virtue of the fact that the GEMRecHit algorithm includes a clusterizer, so GE0 digis are also clustered so we don't get unnecessary duplicates.

**Comment by perrotta** [comment]
+1
- RU segment builder implemented for the GE0 instead of the GEMSegmentAlgorithm, which is used for the other GEM stations. 
- Jenkins tests pass and show no differences (likely because of their low stat) 
- Differences in the number of segments and recHits are visible in some large stat validation shown in the slides attached: they go in the expected direction

**Comment by silviodonato** [comment]
do you have any comments @cms-sw/upgrade-l2 ?

**Comment by srimanob** [comment]
+Upgrade

**Comment by silviodonato** [comment]
+1

**Pull request merged**